File: frontend/src/types/ticket.ts
import z from "zod"

export const ticketFormSchema = z
    .object({
        title: z.string().min(1, "Title is required"),
        description: z.string().min(1, "Description is required"),
        projectId: z.number().min(1, "Project ID is required"),
        sprintId: z.number().min(1, "Sprint ID is required"),
        assigneeId: z.number().optional(),
        issueType: z.enum(["bug", "feature", "task"]),
        priority: z.enum(["low", "medium", "high"]),
        startDate: z.date().optional(),
        endDate: z.date().optional(),
        status: z.enum(["todo", "in_progress", "test", "done"]),
        comments: z.array(z.object({
            content: z.string().min(1, "Comment is required"),
            userId: z.number().min(1, "User ID is required"),
        })),
        attachments: z.array(z.object({
            url: z.string().min(1, "Attachment URL is required"),
            userId: z.number().min(1, "User ID is required"),
        })),
    })
    .refine((data) => data.assigneeId != null && data.assigneeId >= 1, {
        message: "Assignee is required",
        path: ["assigneeId"],
    })

export type TicketForm = z.infer<typeof ticketFormSchema>

File: backend/app/models/ticket.rb
class Ticket < ApplicationRecord
  include BumpsProjectBoardCache

  belongs_to :project
  belongs_to :sprint
  belongs_to :assignee, class_name: "User"

  has_many :comments, dependent: :destroy

  enum :status, { todo: 0, in_progress: 1, test: 2, done: 3 }
  enum :issue_type, { bug: 0, feature: 1, task: 2 }
  enum :priority, { low: 0, medium: 1, high: 2 }

  validates :start_date, presence: true
  validates :end_date, presence: true

  validate :end_date_after_start_date

  private
  
  def end_date_after_start_date
    return if start_date.blank? || end_date.blank?
  end

end
